Heltec SSD1306 Display Support

NorthernMan54 opened the pull request:
## Description:

I needed to rebase my environment against the latest development.  This is a continuation of #1277

This pull request includes

- Support for HELTEC SSD1306 OLED Display as a separate environment 'heltec-display-ssd1306'
- Default log level on LCD display set to WARNING
- Refactored M5Print to a macro based function that is shareable between M5 and Heltec Boards.

Outstanding issue

- Usage with BT Module with log level set to Notice causes exception - https://github.com/1technophile/OpenMQTTGateway/pull/1277#issuecomment-1273551396


## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] I accept the [DCO](https://github.com/1technophile/OpenMQTTGateway/blob/development/docs/participate/development.md#developer-certificate-of-origin).

NorthernMan54 committed:
Mid Removal of unneeded Serial components

NorthernMan54 committed:
Almost complete

My environment

[env:heltec-display]   ; Heltec ESP32 Board with SSD1306 display
platform = ${com.esp32_platform}
board = heltec_wifi_lora_32_V2
build_type = debug
monitor_filters = esp32_exception_decoder
board_build.partitions = min_spiffs.csv
extra_scripts = pre:scripts/omg_firmware_version.py
lib_deps =
  ${com-esp.lib_deps}
  ${libraries.wifimanager32}
  ${libraries.heltec}
;  ${libraries.m5stickcp}

build_flags =
  ${com-esp.build_flags}
  ${test_env.build_flags}
; *** OpenMQTTGateway Config ***
  '-UZmqttDiscovery'          ; disables MQTT Discovery
  '-DvalueAsATopic=true'
  '-DLOG_LEVEL=LOG_LEVEL_TRACE'
; *** OpenMQTTGateway Modules ***
  '-DZboardHELTEC="HELTEC"'
; *** Heltec Board Options ***
  '-DDEBUG_OLEDDISPLAY(...)=Serial.printf( __VA_ARGS__ )'
  '-DLOG_TO_LCD=true'         ; Enable log to LCD
  '-DLED_ERROR=2'
  '-DONBOARD_LED=25'          ; Onboard LED is GPIO 25 on the Heltec Board

monitor_port = /dev/cu.SLAB_USBtoUART
monitor_speed = 921600
upload_port = /dev/cu.SLAB_USBtoUART
upload_speed = 921600

NorthernMan54 committed:
Consolidate files and simplify

NorthernMan54 committed:
Lint fixes

NorthernMan54 committed:
Oops in platformio.ini

NorthernMan54 committed:
This compiles with the LoRa module enabled

Switched to the thingpulse/ESP8266 and ESP32 OLED driver for SSD1306 displays@^4.3.0 library which was the base for the Heltec library.

NorthernMan54 committed:
Lint fixes

NorthernMan54 committed:
Refactored modules and simplified

NorthernMan54 committed:
Lint fixes

NorthernMan54 committed:
Tuning of display buffer

This fixes multi-line display

NorthernMan54 committed:
Implement suggestions

1 - Set default LCD log level to WARNING
2 - Implemented a macro based displayPrint function that will be usable across different displays

NorthernMan54 committed:
Fix conflict

NorthernMan54 committed:
Merge remote-tracking branch 'upstream/development' into Heltec-Display

NorthernMan54 committed:
Updated

NorthernMan54 committed:
Update platform.ini

1technophile commented on main/config_M5.h:
```diff
@@ -65,4 +65,14 @@ extern void loopM5();
+// Simple construct for displaying message in lcd and oled displays
+
+#define displayPrint(...) \
+  if (lowpowermode < 2) return M5Print(__VA_ARGS__) // only print if not in low power mode
```
Could you remove the `return` here, when used the program goes out fo the function and bypass the remaining setup steps, should be `if (lowpowermode < 2) M5Print(__VA_ARGS__)` I think

NorthernMan54 replied:
Makes sense

1technophile commented on main/config_HELTEC.h:
```diff
@@ -0,0 +1,114 @@
+// Simple construct for displaying message in lcd and oled displays
+
+#define displayPrint(...) \
+  if (lowpowermode < 2) return heltecPrint(__VA_ARGS__) // only print if not in low power mode
```
same here for the `return`

NorthernMan54 replied:
Ooops

1technophile commented on main/ZboardHeltec.ino:
```diff
@@ -0,0 +1,225 @@
+#  endif
+}
+
+void loopHELTEC() {
```
In `loopM5()` I redraw the logo if there are no new messages so that the logo is displayed if the gateway is performing normaly.
Do you think we could add this behavior ?

NorthernMan54 replied:
Added this by duplicating the M5 loop function.

NorthernMan54 committed:
Recommended changes

- removed extraneous return from displayPrint
- Added logo display when idle
- Fixed log-lcd mqtt function

NorthernMan54 committed:
Lint fixes

1technophile commented:
Thanks, LGTM
Test OK on the Heltec + BLE
Test OK on M5StickC Plus + BLE

NorthernMan54 commented:
Awesome

NorthernMan54 committed:
Enable OLED Display on Heltec Boards

ssalonen commented:
Did this PR accidentally commit new file, `main.ino.cpp`?

NorthernMan54 commented:
@ssalonen - That is correct, it should not have been included
